plugin/supervisor.go

```diff
@@ -103,6 +110,63 @@ func (sup *supervisor) Hooks() Hooks {
 	return sup.hooks
 }
 
+// PerformHealthCheck checks the plugin through a process check, an RPC ping, and a HealthCheck hook call.
+func (sup *supervisor) PerformHealthCheck() error {
+	if err := sup.CheckProcess(); err != nil {
+		mlog.Debug(fmt.Sprintf("Error checking plugin process, error: %s", err.Error()))
+		return fmt.Errorf("Plugin process not found, or not responding")
+	}
+
+	if err := sup.Ping(); err != nil {
+		for pingFails := 1; pingFails < HEALTH_CHECK_PING_FAIL_LIMIT; pingFails++ {
+			err = sup.Ping()
+			if err == nil {
+				break
+			}
```

Comment:
I'm thinking we should space out the pings a bit and account for timeouts. Something like this:

```go
var pingErr error
for pingFails := 0; pingFails < HEALTH_CHECK_PING_FAIL_LIMIT; pingFails++ {
  c := make(chan error, 1)
  go func() {
    c <- sup.Ping() 
  }()

  select {
  case err := <-c: // Ping() either successful, or explicitly failed before timeout
      if err != nil { // Ping() explicitly failed
          pingErr = err
          time.Sleep(50 * time.Millisecond) // sleep for 50ms
      } else { // Ping() was successful
          break
      }
  case <-time.After(50 * time.Millisecond): // account for 50ms timeout
      pingErr = errors.New("Ping timed out")
  }
}

if pingErr != nil {
  mlog.Debug(fmt.Sprintf("Error pinging plugin, error: %s", pingErr.Error()))
  return fmt.Errorf("Plugin RPC connection is not responding")  
}
```

Reply:
Doesn't Ping timeout itself?

Reply:
Looking at the source for go-plugin and net/rpc, we should be good to not include the timeout check.

How do you feel about including the spaced out ping calls here with `time.Sleep()`?

Reply:
I'm thinking that the timeout isn't appropriate here for Ping, but with the case of the HealthCheck hook, the plugin author may have an issue that causes a never ending block during the hook's execution. We would want to avoid that as I believe it would cause the health check loop to block indefinitely.